### DISCUSSION TITLE
"Virtual" foreign keys

### DISCUSSION THREAD

**Discussion by kasium** [sqlalchemy/sqlalchemy#7075] [Ideas]
Hey guys,

in an existing application I was forced to drop a few foreign keys because of various reasons. I did that with alembic, however, the SQLAlchemy model still assumes the exist, I did some research and it seems, that the replacement should be some relationships.

IMHO, these are not very nice to implement especially if you have existing ones. Also we have code which works with FK metadata which must now be rewritten.

So in short, I want to ask if a kind if "virtual" flag can be added to foreign keys. Meaning, that this code
```python
class User(Base):
    group_id = Column(Integer, ForeignKey(Group.id, virtual=True))

class Group(Base):
    id= Column(Integer, primary_key=True)
    name = Column(String(10))
```
will work as always, but with the exception, that the FK is not created on the database but just exists in the metadata.
This kind of declaration is easier than with relationships and enables all nice features.

Any feedback?

**Comment by CaselIT** [comment]
Hi,

I'm not sure what advantage would it have compared to just removing them. FK are used by sqlalchemy to automatically populate the joins for relationships, but if you are not using relationship I think not much would change.

> enables all nice features.

what are you referring towards?

**Reply by kasium** [reply]
Please correct me if I'm wrong, but the FKs are also used to determine join conditions w/o the need to specify them explicitly. E.g `session.query(A).join(B).all()`

**Reply by CaselIT** [reply]
Yes, I forgot about it. I guess they use the same logic used by relationships.

Is the issue only that alembic wants to autogenerate them in migrations? Or something else

**Reply by kasium** [reply]
Not really. So, let me try to rephrase it. Let's say I have a model with FKs and for whatever reasons I need to remove them from the database.
Now, to make the database and SA model consistent, I need to remove the FKs and replace with relationships. However, relationshipss have a longer declaration and other stuff like lazy loading which I need to take care off.

So instead, I suggest to add a "virtual" or "only_metadata" flag to a FK in SA which tells SA to only create the metadata for this FK but not to push it to the database.

With that, I recreation of the model will not create them and a code reader is aware, that he/she can't expect, that the FKs exist in the database.

So in summary, I want to have a relationship in the sytax of a FK, because the FK syntax is more straightforward and easier to use.

**Reply by CaselIT** [reply]
@zzzeek I'm guessing something similar is not supported at the moment.

If we consider this, it would also require alembic changes

**Reply by kasium** [reply]
@CaselIT should I create a new issue from this discussion, since I guess this is a valid feature request? What do you think?

**Reply by CaselIT** [reply]
@zzzeek what's your opinion on this?

**Reply by zzzeek** [reply]
I dont see what the problem using primaryjoin with foreign() in relationship() is. if this is just a "few" foreign keys, it does not seem to be a burden to have the relationships include an extra parameter.

OTOH, if one is looking to toggle on and off *all* foreign keys, like staging vs. production, then it seems burdensome that every individual FK would have a "virtual" flag that needs to be flipped, a schema-wide solution would be better.  Back in the Oracle days we used to do this using some Oracle directive which accomplishes this in fact.

so we have "all" vs. a "few".  the proposal doesn't work well for "all" and it seems excessive for "few" (we *really* do not like to add flags to things if there are other ways to accomplish something.   note how overwhelming our API is to people already).

if you've dropped the FKs from the Alembic migrations, I would say you're done, keep the ForeignKey defs in your model metadata.

if the problem is autogenerate, then here's what you do:

1. put a flag in the info area of ForeignKey:

```py
class User(Base):
    group_id = Column(Integer, ForeignKey(Group.id, info={"omit_from_schema": True}))
```

2. add a [include_object](https://alembic.sqlalchemy.org/en/latest/api/runtime.html?highlight=include_object#alembic.runtime.environment.EnvironmentContext.configure.params.include_object) function to omit:

```py
def include_object(object, name, type_, reflected, compare_to):
    if (type_ == "foreign_key_constraint" and
        not reflected and
        object.info.get("omit_from_schema", False)):
        return False
    else:
        return True

context.configure(
    # ...
    include_object = include_object
)
```

3. add an alternate render for ForeignKeyConstraint so it doesnt render for plain create table

```py
@compiles(ForeignKeyConstraint)
def skip_fk(element, compiler, **kw):
    for fk in element.elements:
        if fk.info.get("omit_from_schema", False):
            return None
    return compiler.visit_foreign_key_constraint(element, **kw)


class User(Base):
    __tablename__ = "u"
    id = Column(Integer, primary_key=True)
    group_id = Column(
        Integer, ForeignKey("g.id", info={"omit_from_schema": True})
    )

```


i will gladly add this to alembic recipes

**Reply by kasium** [reply]
@zzzeek thanks a lot. Very valid arguments and good solution for alembic. So, I'll stay with the ForeignKey declarations for now.

**Reply by kasium** [reply]
Not sure how to flag the discussion as done, so if you know how to close it feel free to do so
